<a href="https://colab.research.google.com/github/akashde1998-Alpha/GCN-by-pytorch-/blob/main/Dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [48]:
!pip install torch-geometric



### Cell Description
This cell installs the `torch-geometric` library, which is essential for working with graph neural networks in PyTorch.

In [19]:
from torch_geometric.datasets import Planetoid
dataset=Planetoid(root='data/Cora',name='Cora')
data=dataset[0]

### Cell Description
This cell imports the `Planetoid` dataset from `torch_geometric.datasets` and loads the Cora dataset. The Cora dataset is a classic graph dataset used for node classification tasks. `data = dataset[0]` extracts the first (and typically only) graph from the dataset object.

In [3]:
print(data)

Data(x=[2708, 1433], edge_index=[2, 10556], y=[2708], train_mask=[2708], val_mask=[2708], test_mask=[2708])


### Cell Description
This cell prints the `data` object, which represents the loaded Cora graph. The output `Data(x=[2708, 1433], edge_index=[2, 10556], y=[2708], train_mask=[2708], val_mask=[2708], test_mask=[2708])` shows its key attributes:
- `x`: Node feature matrix (2708 nodes, 1433 features per node).
- `edge_index`: Graph connectivity in COO format (2 rows for source/target nodes, 10556 edges).
- `y`: Node labels/targets (2708 labels).
- `train_mask`, `val_mask`, `test_mask`: Boolean masks indicating which nodes belong to the training, validation, and test sets respectively.

In [4]:
data.x[37]

tensor([0., 0., 0.,  ..., 0., 0., 0.])

### Cell Description
This cell accesses and displays the feature vector for node at index 37 from the `data.x` tensor. The output `tensor([0., 0., 0., ..., 0., 0., 0.])` shows a sparse representation of its features, indicating the presence or absence of specific words from the vocabulary.

In [5]:
for i in range(2708):
  if data.x[i,0]==1 and data.x[i,2]==1:
    print(data.x[i],i)
  else:
    pass

tensor([1., 0., 1.,  ..., 0., 0., 0.]) 793


### Cell Description
This cell iterates through all 2708 nodes (`i` from 0 to 2707). For each node, it checks if both the feature at index 0 and the feature at index 2 are present (i.e., equal to 1). If both conditions are met, it prints the full feature vector of that node and its index `i`. This helps in finding nodes that share specific common features.

In [20]:
#The
for j in range(1433):
  if data.x[0][j] == 1:
    print(j)

19
81
146
315
774
877
1194
1247
1274


### Cell Description
This cell iterates through all 1433 possible features (`j` from 0 to 1432) for the *first node* (node 0). It checks if a specific feature `j` is present (i.e., `data.x[0][j] == 1`). If a feature is present, its index `j` is printed. This effectively shows the indices of the words present in the bag-of-words representation for node 0.

In [47]:
 data.edge_index


tensor([[ 633, 1862, 2582,  ...,  598, 1473, 2706],
        [   0,    0,    0,  ..., 2707, 2707, 2707]])

### Cell Description
This cell directly displays the `data.edge_index` tensor. This tensor represents the graph's adjacency list in COO (Coordinate) format. The first row lists the source nodes, and the second row lists the target nodes. Each column `[u, v]` signifies an edge from node `u` to node `v`.

In [44]:
data.edge_index[0, :].unsqueeze(0)

tensor([[ 633, 1862, 2582,  ...,  598, 1473, 2706]])

### Cell Description
This cell extracts the first row of the `data.edge_index` tensor (i.e., `data.edge_index[0, :]`), which contains all the *source* nodes of the edges. It then uses `.unsqueeze(0)` to ensure the result is a row vector (a 2D tensor with one row), which is often useful for tensor operations that expect specific dimensions.

In [42]:
data.edge_index[:, 0].unsqueeze(1)


tensor([[633],
        [  0]])

### Cell Description
This cell extracts the first column of the `data.edge_index` tensor (i.e., `data.edge_index[:, 0]`), which represents the source and target nodes of the *first edge* in the graph. It then uses `.unsqueeze(1)` to transform this 1D tensor into a column vector (a 2D tensor with one column), which was specifically requested earlier by the user to visualize it in a column format.